### PR TITLE
fix(consolidation): superseded facts no longer re-promoted from unchanged episodic sources

### LINKED ISSUE #227: consolidation: superseded facts can be re-promoted from unchanged episodic sources
Found during review of #219 (merged): the idempotency guard seeds already_promoted from get_all_memories_by_namespace, which excludes superseded rows (by design since the supersession primitive, commit 7150cc8). Superseding a consolidation-produced semantic "mentioned" fact therefore removes it from the guard set while its source episodic memories remain intact — the next consolidation run re-derives the identical (entity, content) pair and writes a fresh, active semantic row, reinstating content that was just corrected. One-time resurrection per correction (the new row then re-enters the guard set), so growth is bounded, but it undoes the supersession.

The intended re-assertion pathway (new episodic evidence re-establishing a fact after invalidation) is different from this case (re-derivation from the same old evidence). Candidate fix: seed the guard from get_all_memories_by_namespace_including_superseded (or union in superseded "mentioned" rows) so a superseded key still suppresses re-derivation, while genuinely new episodic evidence — which changes the cluster — can still promote. Needs a test covering the supersede-then-consolidate sequence; none exists today.

### PR DESCRIPTION
**Pull request by wshobson**
## The bug

`promote_episodic_to_semantic` seeds its idempotency guard, keyed on `(about_entity, content)`, from `get_all_memories_by_namespace` — which excludes superseded rows by design (supersession primitive, #200).

So when a consolidation-produced semantic `mentioned` fact was superseded — i.e. someone corrected it — its key dropped out of the guard set, while the episodic memories it was derived from stayed intact and active. The next consolidation run re-derived the identical `(entity, content)` pair from that unchanged evidence and wrote a fresh **ACTIVE** semantic row. The correction was silently undone, and because `pensyve_episode_end` spawns a namespace consolidation, this recurred every session.

## The fix

Read the namespace with `get_all_memories_by_namespace_including_superseded` so superseded promotions are visible to the guard, and record *why* each key is closed rather than just that it is:

- **`Active`** — a live row holds the key. Closed unconditionally; this is the pre-existing duplicate suppression.
- **`SupersededAt(t)`** — every row holding the key is superseded, `t` being the supersession moment stamped on the row by the supersession primitive (`invalid_at`, the third argument to `supersede_memory`). Closed only against clusters whose episodes *all* predate `t`.

That scoping is what keeps #227's two cases apart. A correction speaks to the evidence that existed when it was made, so re-deriving the key from episodes older than it would undo the correction; a cluster carrying at least one episode recorded *after* it is new testimony and promotes — **including when it re-asserts the very content that was corrected**. Keying superseded rows forever would have tombstoned the entity on that subject permanently.

Where several rows share a key, a live row wins and otherwise the latest supersession does, since that is the correction still standing. A supersession with no stamped moment leaves nothing to date evidence against, so it keeps closing the key outright.

Because the history read also returns superseded *episodic* rows, the partition sweep skips those: retired evidence must not seed clusters, which keeps clustering behaving exactly as before. The sweep moved into `partition_for_promotion` to keep the promotion pass within its line budget.

No new storage method was needed; the trait already declared `get_all_memories_by_namespace_including_superseded`.

## Tests

New file `pensyve-core/tests/test_consolidation_supersession_guard.rs`:

1. `superseded_promotion_is_not_reminted_from_unchanged_evidence` — the regression test. Promote, supersede with a correction, consolidate twice more. Asserts no new promotions, that the only active `mentioned` row is the correction, and that the original row is still superseded.
2. `new_evidence_reasserting_same_content_promotes_after_supersession` — the re-assertion pathway with **identical** content: "dark mode" promoted → corrected to "light mode" → fresh episodes say "dark mode" again 60 days later → promotes, and does not duplicate on the run after that.
3. `new_evidence_still_promotes_after_supersession` — re-assertion with different content.
4. `superseded_episodic_evidence_cannot_support_a_cluster` — a retired episode cannot make up the second member of a cluster, while live evidence beside it stays eligible.

Tests 1 and 3 were written first and failed against the unfixed code:

```
---- superseded_promotion_is_not_reminted_from_unchanged_evidence stdout ----
assertion `left == right` failed: run 2 resurrected a superseded fact
  left: 1
 right: 0

---- new_evidence_still_promotes_after_supersession stdout ----
assertion `left == right` failed: new episodic evidence must still promote after a supersession
  left: 2
 right: 1
```

(The second is the same bug from the other side: the run promoted the new fact *plus* the resurrected superseded one.)

Test 2 was then written against the first commit's tombstone guard and failed there:

```
---- new_evidence_reasserting_same_content_promotes_after_supersession stdout ----
assertion `left == right` failed: episodes recorded after the correction must re-assert the fact
  left: 0
 right: 1
```

Supersession tests pin explicit timestamps, so the ordering they depend on is deterministic rather than wall-clock luck.

## Verification

- `cargo test -p pensyve-core` — 640 passed, 0 failed, 9 ignored
- `cargo test -p pensyve-core --features postgres` — 644 passed, 0 failed, 9 ignored
- `cargo fmt --check -p pensyve-core` — clean
- `cargo clippy -p pensyve-core --all-targets` — clean
- The three pre-existing tests in `test_consolidation_idempotency.rs` remain green.

Closes #227

https://claude.ai/code/session_01AsVz67dYuk3dufxA683hJA

### CODE REVIEW

File: pensyve-core/src/consolidation/mod.rs [hunk 1]
```diff
@@ -232,12 +236,15 @@ impl ConsolidationEngine {
...
+                // new clusters; only live rows feed the clustering pass.
+                Memory::Episodic(em) if em.superseded_by.is_none() => {
                     episodic_by_entity
                         .entry(em.about_entity)
                         .or_default()
                         .push(em);
                 }
+                // Both live and superseded promotions seed the guard.
                 Memory::Semantic(sm) if sm.predicate == "mentioned" => {
                     already_promoted.insert((sm.subject, sm.object));
```
[the remark this answers is not in the document]
Comment (wshobson) [reply pensyve-core/src/consolidation/mod.rs]: Addressed in e09a2ed.

You were right that the first commit made the guard a permanent tombstone, which contradicts the semantics #227 asks for: new episodic evidence must be able to re-assert a fact after invalidation, including the same content that was corrected.

The flat `HashSet<(entity, content)>` is now a `HashMap<(entity, content), PromotionGuard>`:

- `Active` — a live row holds the key; closed unconditionally (duplicate suppression, unchanged behavior).
- `SupersededAt(t)` — every row holding the key is superseded; `t` is the supersession moment stamped on the row by the supersession primitive (`invalid_at`, the third argument to `supersede_memory`). The key is closed only against clusters whose episodes all predate `t`. A cluster carrying at least one episode recorded after `t` is new testimony and promotes.

Where several rows share a key, a live row wins and otherwise the latest supersession does, since that is the correction still standing. A supersession with no stamped moment leaves nothing to date evidence against, so it keeps closing the key outright.

New test `new_evidence_reasserting_same_content_promotes_after_supersession` walks exactly the case from the review — promoted "dark mode" → superseded by "light mode" → fresh episodes say "dark mode" again 60 days later → promotes, and does not duplicate on the run after that. It failed on the previous commit (`left: 0, right: 1`). The old-evidence-only regression test stays green.

Separately, `superseded_episodic_evidence_cannot_support_a_cluster` covers the `Memory::Episodic(em) if em.superseded_by.is_none()` arm that CodeRabbit flagged as untested: a retired episode cannot make up the second member of a cluster, while live evidence beside it stays eligible.

Verification: `cargo test -p pensyve-core` 640 passed / 0 failed; `--features postgres` 644 passed / 0 failed; `cargo fmt --check` and `cargo clippy -p pensyve-core --all-targets` clean.

[the remark this answers is not in the document]
Comment (wshobson) [reply pensyve-core/src/consolidation/mod.rs]: Addressed in e09a2ed.

You were right that the first commit made the guard a permanent tombstone, which contradicts the semantics #227 asks for: new episodic evidence must be able to re-assert a fact after invalidation, including the same content that was corrected.

The flat `HashSet<(entity, content)>` is now a `HashMap<(entity, content), PromotionGuard>`:

- `Active` — a live row holds the key; closed unconditionally (duplicate suppression, unchanged behavior).
- `SupersededAt(t)` — every row holding the key is superseded; `t` is the supersession moment stamped on the row by the supersession primitive (`invalid_at`, the third argument to `supersede_memory`). The key is closed only against clusters whose episodes all predate `t`. A cluster carrying at least one episode recorded after `t` is new testimony and promotes.

Where several rows share a key, a live row wins and otherwise the latest supersession does, since that is the correction still standing. A supersession with no stamped moment leaves nothing to date evidence against, so it keeps closing the key outright.

New test `new_evidence_reasserting_same_content_promotes_after_supersession` walks exactly the case from the review — promoted "dark mode" → superseded by "light mode" → fresh episodes say "dark mode" again 60 days later → promotes, and does not duplicate on the run after that. It failed on the previous commit (`left: 0, right: 1`). The old-evidence-only regression test stays green.

Separately, `superseded_episodic_evidence_cannot_support_a_cluster` covers the `Memory::Episodic(em) if em.superseded_by.is_none()` arm that CodeRabbit flagged as untested: a retired episode cannot make up the second member of a cluster, while live evidence beside it stays eligible.

Verification: `cargo test -p pensyve-core` 640 passed / 0 failed; `--features postgres` 644 passed / 0 failed; `cargo fmt --check` and `cargo clippy -p pensyve-core --all-targets` clean.

**Comment by wshobson** [comment]
Addressed in e09a2ed.

You were right that the first commit made the guard a permanent tombstone, which contradicts the semantics #227 asks for: new episodic evidence must be able to re-assert a fact after invalidation, including the same content that was corrected.

The flat `HashSet<(entity, content)>` is now a `HashMap<(entity, content), PromotionGuard>`:

- `Active` — a live row holds the key; closed unconditionally (duplicate suppression, unchanged behavior).
- `SupersededAt(t)` — every row holding the key is superseded; `t` is the supersession moment stamped on the row by the supersession primitive (`invalid_at`, the third argument to `supersede_memory`). The key is closed only against clusters whose episodes all predate `t`. A cluster carrying at least one episode recorded after `t` is new testimony and promotes.

Where several rows share a key, a live row wins and otherwise the latest supersession does, since that is the correction still standing. A supersession with no stamped moment leaves nothing to date evidence against, so it keeps closing the key outright.

New test `new_evidence_reasserting_same_content_promotes_after_supersession` walks exactly the case from the review — promoted "dark mode" → superseded by "light mode" → fresh episodes say "dark mode" again 60 days later → promotes, and does not duplicate on the run after that. It failed on the previous commit (`left: 0, right: 1`). The old-evidence-only regression test stays green.

Separately, `superseded_episodic_evidence_cannot_support_a_cluster` covers the `Memory::Episodic(em) if em.superseded_by.is_none()` arm that CodeRabbit flagged as untested: a retired episode cannot make up the second member of a cluster, while live evidence beside it stays eligible.

Verification: `cargo test -p pensyve-core` 640 passed / 0 failed; `--features postgres` 644 passed / 0 failed; `cargo fmt --check` and `cargo clippy -p pensyve-core --all-targets` clean.

**Comment by wshobson** [comment]
@coderabbitai review

**Comment by wshobson** [comment]
@coderabbitai review

**Comment by wshobson** [comment]
@coderabbitai review
